feat: Tmux control mode

joexue opened the pull request:
Complete the unfinished tmux control mode functionalities, include:

- Sync the windows/panels/layouts from back end.
- Split panels
- Add windows
- Prune the panels/windows when the back end quit them 
- Change the panels size
- Change the window size

Passed all tests by "cargo test --all"
Updated the "cargo fmt --all"

![Screenshot 2025-01-16 205600](https://github.com/user-attachments/assets/b4439dc3-cdb2-48ae-bd77-fa7f7efc7258)

jayPare commented:
Didn't looked at the code since I'm not very familiar with Wezterm code base, but I might take a quick look. However, I built from source on Windows, and everything seems to work properly. Thank you for that!

joexue commented:
> Didn't looked at the code since I'm not very familiar with Wezterm code base, but I might take a quick look. However, I built from source on Windows and tested connecting to tmux 3.4 on a Linux machine. Everything seems to work properly. Thank you for that!

Thanks for testing it, will push the last part tonight, then I think all functionalities are done.

felixding commented:
Thank you for your work. AFAIK, this is the first full implementation of tmux control mode besides iTem2. Really looking forward to it.

joexue commented:
All are done, welcome and please review and comment.

bfstree commented:
Thank you for the commits! However, I get an immediate crash after typing `tmux -CC` using tmux version 3.5a.

```
16:02:11.053  INFO   mux::localpane > tmux -CC mode requested
16:02:11.053  INFO   mux::tmux      > tmux: Guarded(Guarded { error: false, timestamp: 1739048531, number: 314, flags: 0, output: "" }) in state WaitForInitialGuard
16:02:11.053  INFO   mux::tmux      > tmux: WindowAdd { window: 2 } in state Idle
16:02:11.053  INFO   mux::tmux      > Tmux create window id 1
16:02:11.053  INFO   mux::tmux      > tmux: SessionsChanged in state Idle
16:02:11.053  INFO   mux::tmux      > tmux: SessionChanged { session: 2, name: "2" } in state Idle
16:02:11.053  INFO   mux::tmux      > tmux session changed:2
16:02:11.054  INFO   mux::tmux      > sending cmd "list-panes -aF '#{session_id} #{window_id} #{pane_id} #{pane_index} #{cursor_x} #{cursor_y} #{pane_width} #{pane_height} #{pane_left} #{pane_top}'\n"
16:02:11.058  INFO   mux::tmux      > tmux: Guarded(Guarded { error: false, timestamp: 1739048531, number: 320, flags: 1, output: "$0 @0 %0 0 18 0 80 24 0 0\n$1 @1 %1 0 18 0 80 24 0 0\n$2 @2 %2 0 0 0 80 24 0 0\n" }) in state WaitingForResponse
16:02:11.091  ERROR  env_bootstrap  > panic at /Users/*******/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.1/src/runnable.rs:451:17 - local task polled by a thread that didn't spawn it
   0: backtrace::capture::Backtrace::create
   1: backtrace::capture::Backtrace::new
   2: env_bootstrap::register_panic_hook::{{closure}}
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic_handler::{{closure}}
   5: std::sys::backtrace::__rust_end_short_backtrace
   6: _rust_begin_unwind
   7: core::panicking::panic_fmt
   8: <async_task::runnable::Builder<M>::spawn_local::Checked<F> as core::future::future::Future>::poll
   9: async_task::raw::RawTask<F,T,S,M>::run
  10: window::spawn::SpawnQueue::run
  11: window::spawn::SpawnQueue::trigger
  12: <unknown>
  13: <unknown>
  14: <unknown>
  15: <unknown>
  16: <unknown>
  17: <unknown>
  18: <unknown>
  19: <unknown>
  20: <unknown>
  21: <unknown>
  22: <window::os::macos::connection::Connection as window::connection::ConnectionOps>::run_message_loop
  23: wezterm_gui::run_terminal_gui
  24: wezterm_gui::main
  25: std::sys::backtrace::__rust_begin_short_backtrace
  26: std::rt::lang_start::{{closure}}
  27: std::rt::lang_start_internal
  28: _main
```

jayPare commented:
> Thank you for the commits! However, I get an immediate crash after typing `tmux -CC` using tmux version 3.5a.
> 
> ```
> 16:02:11.053  INFO   mux::localpane > tmux -CC mode requested
> 16:02:11.053  INFO   mux::tmux      > tmux: Guarded(Guarded { error: false, timestamp: 1739048531, number: 314, flags: 0, output: "" }) in state WaitForInitialGuard
> 16:02:11.053  INFO   mux::tmux      > tmux: WindowAdd { window: 2 } in state Idle
> 16:02:11.053  INFO   mux::tmux      > Tmux create window id 1
> 16:02:11.053  INFO   mux::tmux      > tmux: SessionsChanged in state Idle
> 16:02:11.053  INFO   mux::tmux      > tmux: SessionChanged { session: 2, name: "2" } in state Idle
> 16:02:11.053  INFO   mux::tmux      > tmux session changed:2
> 16:02:11.054  INFO   mux::tmux      > sending cmd "list-panes -aF '#{session_id} #{window_id} #{pane_id} #{pane_index} #{cursor_x} #{cursor_y} #{pane_width} #{pane_height} #{pane_left} #{pane_top}'\n"
> 16:02:11.058  INFO   mux::tmux      > tmux: Guarded(Guarded { error: false, timestamp: 1739048531, number: 320, flags: 1, output: "$0 @0 %0 0 18 0 80 24 0 0\n$1 @1 %1 0 18 0 80 24 0 0\n$2 @2 %2 0 0 0 80 24 0 0\n" }) in state WaitingForResponse
> 16:02:11.091  ERROR  env_bootstrap  > panic at /Users/*******/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.1/src/runnable.rs:451:17 - local task polled by a thread that didn't spawn it
>    0: backtrace::capture::Backtrace::create
>    1: backtrace::capture::Backtrace::new
>    2: env_bootstrap::register_panic_hook::{{closure}}
>    3: std::panicking::rust_panic_with_hook
>    4: std::panicking::begin_panic_handler::{{closure}}
>    5: std::sys::backtrace::__rust_end_short_backtrace
>    6: _rust_begin_unwind
>    7: core::panicking::panic_fmt
>    8: <async_task::runnable::Builder<M>::spawn_local::Checked<F> as core::future::future::Future>::poll
>    9: async_task::raw::RawTask<F,T,S,M>::run
>   10: window::spawn::SpawnQueue::run
>   11: window::spawn::SpawnQueue::trigger
>   12: <unknown>
>   13: <unknown>
>   14: <unknown>
>   15: <unknown>
>   16: <unknown>
>   17: <unknown>
>   18: <unknown>
>   19: <unknown>
>   20: <unknown>
>   21: <unknown>
>   22: <window::os::macos::connection::Connection as window::connection::ConnectionOps>::run_message_loop
>   23: wezterm_gui::run_terminal_gui
>   24: wezterm_gui::main
>   25: std::sys::backtrace::__rust_begin_short_backtrace
>   26: std::rt::lang_start::{{closure}}
>   27: std::rt::lang_start_internal
>   28: _main
> ```

Are you on Windows or Linux? Also on which version might help!

bfstree commented:
> Are you on Windows or Linux? Also on which version might help!

I'm on macOS 15.3

fabian-thomas commented:
I observe the same crash on Linux on bc4d4ba5eb03f698253972947b4cbbaea2fddec5. Maybe the issue is how we start tmux? Do I understand it correctly that I I have to launch wezterm like this? `wezterm ssh machine -- tmux -CC`
Nevertheless, thanks for your work on this. Waiting for something like this since years.

joexue commented:
> I observe the same crash on Linux on [bc4d4ba](https://github.com/wezterm/wezterm/commit/bc4d4ba5eb03f698253972947b4cbbaea2fddec5). Maybe the issue is how we start tmux? Do I understand it correctly that I I have to launch wezterm like this? `wezterm ssh machine -- tmux -CC` Nevertheless, thanks for your work on this. Waiting for something like this since years.

Not sure if there is problem when you mixed ssh domain and tmux domain, I don’t have environment to test tat for now. could you just ssh to machine then run “ tmux -CC” and see what happene? And what is your tmux version?

joexue commented:
> Thank you for the commits! However, I get an immediate crash after typing `tmux -CC` using tmux version 3.5a.
> 
> ```
> 16:02:11.053  INFO   mux::localpane > tmux -CC mode requested
> 16:02:11.053  INFO   mux::tmux      > tmux: Guarded(Guarded { error: false, timestamp: 1739048531, number: 314, flags: 0, output: "" }) in state WaitForInitialGuard
> 16:02:11.053  INFO   mux::tmux      > tmux: WindowAdd { window: 2 } in state Idle
> 16:02:11.053  INFO   mux::tmux      > Tmux create window id 1
> 16:02:11.053  INFO   mux::tmux      > tmux: SessionsChanged in state Idle
> 16:02:11.053  INFO   mux::tmux      > tmux: SessionChanged { session: 2, name: "2" } in state Idle
> 16:02:11.053  INFO   mux::tmux      > tmux session changed:2
> 16:02:11.054  INFO   mux::tmux      > sending cmd "list-panes -aF '#{session_id} #{window_id} #{pane_id} #{pane_index} #{cursor_x} #{cursor_y} #{pane_width} #{pane_height} #{pane_left} #{pane_top}'\n"
> 16:02:11.058  INFO   mux::tmux      > tmux: Guarded(Guarded { error: false, timestamp: 1739048531, number: 320, flags: 1, output: "$0 @0 %0 0 18 0 80 24 0 0\n$1 @1 %1 0 18 0 80 24 0 0\n$2 @2 %2 0 0 0 80 24 0 0\n" }) in state WaitingForResponse
> 16:02:11.091  ERROR  env_bootstrap  > panic at /Users/*******/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.1/src/runnable.rs:451:17 - local task polled by a thread that didn't spawn it
>    0: backtrace::capture::Backtrace::create
>    1: backtrace::capture::Backtrace::new
>    2: env_bootstrap::register_panic_hook::{{closure}}
>    3: std::panicking::rust_panic_with_hook
>    4: std::panicking::begin_panic_handler::{{closure}}
>    5: std::sys::backtrace::__rust_end_short_backtrace
>    6: _rust_begin_unwind
>    7: core::panicking::panic_fmt
>    8: <async_task::runnable::Builder<M>::spawn_local::Checked<F> as core::future::future::Future>::poll
>    9: async_task::raw::RawTask<F,T,S,M>::run
>   10: window::spawn::SpawnQueue::run
>   11: window::spawn::SpawnQueue::trigger
>   12: <unknown>
>   13: <unknown>
>   14: <unknown>
>   15: <unknown>
>   16: <unknown>
>   17: <unknown>
>   18: <unknown>
>   19: <unknown>
>   20: <unknown>
>   21: <unknown>
>   22: <window::os::macos::connection::Connection as window::connection::ConnectionOps>::run_message_loop
>   23: wezterm_gui::run_terminal_gui
>   24: wezterm_gui::main
>   25: std::sys::backtrace::__rust_begin_short_backtrace
>   26: std::rt::lang_start::{{closure}}
>   27: std::rt::lang_start_internal
>   28: _main
> ```

I tested on tmux 3.5 and did not see the issue. could be something else, I don't have macOS for now unfortunately. And form the log I see you are not using the last commit, could you try that?

bfstree commented:
> I tested on tmux 3.5 and did not see the issue. could be something else, I don't have macOS for now unfortunately. And form the log I see you are not using the last commit, could you try that?

Deleted the entire wezterm repo and re-cloned with your branch. It is now working using tmux 3.5a.
I first `ssh`ed into the server (e.g., `ssh servername`) in wezterm, and both commands `tmux -CC` and `tmux -CC a` work fine. Thanks for your work!

fabian-thomas commented:
> Not sure if there is problem when you mixed ssh domain and tmux domain, I don’t have environment to test tat for now. could you just ssh to machine then run “ tmux -CC” and see what happene? And what is your tmux version?

Ok nope, it works perfectly fine. I just messed up my nix command and installed the wrong branch. Sorry!

joexue commented:
> > Not sure if there is problem when you mixed ssh domain and tmux domain, I don’t have environment to test tat for now. could you just ssh to machine then run “ tmux -CC” and see what happene? And what is your tmux version?
> 
> Ok nope, it works perfectly fine. I just messed up my nix command and installed the wrong branch. Sorry!

Thanks for clarifying

ron-wolf commented:
I want to help test the Windows build, can I do so without an SSH server by just running `tmux -CC`?

joexue commented:
> I want to help test the Windows build, can I do so without an SSH server by just running `tmux -CC`?

Do you have WSL? You can use wsl with tmux installed

ron-wolf commented:
I have WSL installed, but when I build and run for Windows, executing `tmux -CC` in WSL seems to keep that tab busy doing nothing.

iDarshan commented:
I built this on Mac OS 15.3 and tested control mode with `tmux -CC` on ssh server. Its working pretty well. Thanks for adding  control mode!

joexue commented:
> I have WSL installed, but when I build and run for Windows, executing `tmux -CC` in WSL seems to keep that tab busy doing nothing.

I found this is a infra issue, not related to this commit. I could work on it but doesn't block this in.

joexue commented:
> I have WSL installed, but when I build and run for Windows, executing `tmux -CC` in WSL seems to keep that tab busy doing nothing.

@jayPare do you have idea about this? The log shows, the tmux control sequence "\033P1000p" is not send to wezterm, I guess it because of the ENV/terminal setup. I know you successfully run the windows version, how did you start the wezterm?

joexue commented:
> > I found this is a infra issue, not related to this commit. I could work on it but doesn't block this in.
> 
> For the record, I didn’t mean to block the release of this feature, just to help test it across platforms. If it’s prohibitive to fix for Windows, I wouldn’t be opposed to platform-locking it.

I dig it a bit, it is windows ConPTY issue, use wezterm ssh user@host can bypass it but has a stackoverflow issue which is not caused by this commit either. I will create issues for them.

bew commented:
@joexue What do you want to do with those extra issues, do you want me to assign you so you'll work them after this one is merged?

joexue commented:
> @joexue What do you want to do with those extra issues, do you want me to assign you so you'll work them after this one is merged?

Sure, assign them to me. And any tmux cc related issues you can assign me and @me please.

LeszekSwirski commented:
I'm observing this as working on Linux, but I seem to have issues with my Home/End keys not working -- not sure if they're being dropped by wezterm, tmux, or somewhere in between.

pescobar commented:
thanks @joexue for working in this long waited feature :)

I have tested it on linux and control mode is working fine in the light test I did but I noticed that if I launch `tmux` inside wezterm (no control mode) and use right click with mouse I get the tmux menu as in the screenshot below but I start tmux in control mode with `tmux -CC` I don't get the same menu when using right click

![image](https://github.com/user-attachments/assets/90290891-231f-4fc5-88a7-1a829a73909a)

I tested it with tmux 3.2a in linux executing both westerm and tmux in local (no ssh to remote machine)

Is this expected? Does it happen to anyone else?

joexue commented:
> thanks @joexue for working in this long waited feature :)
> 
> I have tested it on linux and control mode is working fine in the light test I did but I noticed that if I launch `tmux` inside wezterm (no control mode) and use right click with mouse I get the tmux menu as in the screenshot below but I start tmux in control mode with `tmux -CC` I don't get the same menu when using right click
> 
> ![image](https://private-user-images.githubusercontent.com/103797/412943505-90290891-231f-4fc5-88a7-1a829a73909a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NTk2OTAsIm5iZiI6MTczOTQ1OTM5MCwicGF0aCI6Ii8xMDM3OTcvNDEyOTQzNTA1LTkwMjkwODkxLTIzMWYtNGZjNS04OGE3LTFhODI5YTczOTA5YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxM1QxNTA5NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMzQ2ZDI4ZjI1OGYyMTU5ODhlNzVjNjFiOWNjOTNlZjg3YTQ3M2E5ZmMxYTI3ODEyNTc3YjQ3ZWFkMDRlZjk4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._MwhThRB_eSiJ9VkJ1FyJfy-mvYz8R6exM8o84zQUT8)
> 
> I tested it with tmux 3.2a in linux executing both westerm and tmux in local (no ssh to remote machine)
> 
> Is this expected? Does it happen to anyone else?

if the menu is from tmux itself, then makes scenes, the point of tmux control mode is to use local renderer/actions to replace tmux's. if you want to do those split actions, use local split command/shortcut/ to do that. the tmux mouse hooks are not mapped to local.

wez commented on mux/src/tmux.rs:
```diff
@@ -206,6 +329,32 @@ impl TmuxDomain {
+        _command_dir: Option<String>,
+        _window: WindowId,
+    ) -> anyhow::Result<Arc<Tab>> {
+        self.inner.create_tmux_window();
```
if we're error'ing out unconditionally on the next line, should this line be here?
If so, please add comments in the code to explain, because it looks redundant!

joexue replied:
yes, the is to send the command to tmux to create a new window, since we don't have the remote window information to return, so we error'ing it. we handle it asynchronously when the window add event come.

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -32,22 +33,46 @@ pub(crate) struct PaneItem {
-            Some(x) => x,
+    pub fn check_pane_attached(&self, window_id: TmuxWindowId, pane_id: TmuxPaneId) -> bool {
+        let gui_tabs = self.gui_tabs.lock();
+        let local_tab = gui_tabs.get(&window_id).unwrap();
```
bare unwrap here; can we eliminate it to avoid crashing wezterm?

```suggestion
        let Some(local_tab) = gui_tabs.get(&window_id) else { return false; };
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -32,22 +33,46 @@ pub(crate) struct PaneItem {
+    pub fn check_pane_attached(&self, window_id: TmuxWindowId, pane_id: TmuxPaneId) -> bool {
+        let gui_tabs = self.gui_tabs.lock();
+        let local_tab = gui_tabs.get(&window_id).unwrap();
+        match local_tab.panes.get(&pane_id) {
+            Some(_) => {
+                return true;
+            }
```
I think the rest of this function can simplify to this:

```suggestion
        local_tab.panes.get(&pane_id).is_some()
```

(Github won't let me include the rest in this suggestion because of the gap in the diff hunk)

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -32,22 +33,46 @@ pub(crate) struct PaneItem {
+
+    pub fn check_window_attached(&self, window_id: TmuxWindowId) -> bool {
+        let gui_tabs = self.gui_tabs.lock();
+        match gui_tabs.get(&window_id) {
```
same sort of comment as for `check_pane_attached`:

```suggestion
        gui_tabs.get(&window_id).is_some()
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
 
+    fn add_attached_window(&self, target: &WindowItem, tab_id: &TabId) -> anyhow::Result<()> {
+        let mut gui_tabs = self.gui_tabs.lock();
+        match gui_tabs.get(&target.window_id) {
+            Some(_x) => return Ok(()),
+            None => {
+                gui_tabs.insert(
+                    target.window_id,
+                    TmuxTab {
+                        tab_id: *tab_id,
+                        tmux_window_id: target.window_id,
+                        layout_csum: target.layout_csum.clone(),
+                        panes: HashSet::new(),
+                    },
+                );
+            }
+        };
+
+        return Ok(());
```
```suggestion
        if !gui_tabs.contains_key(&target.window_id) {
                gui_tabs.insert(
                    target.window_id,
                    TmuxTab {
                        tab_id: *tab_id,
                        tmux_window_id: target.window_id,
                        layout_csum: target.layout_csum.clone(),
                        panes: HashSet::new(),
                    },
                );
        }
        Ok(())
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
+                local_pane.set_cursor_position(pane.cursor_x as usize, pane.cursor_y as usize);
+                if pane.pane_active {
+                    let gui_tabs = self.gui_tabs.lock();
+                    let local_tab = gui_tabs.get(&pane.window_id).unwrap();
```
bare unwrap here; let's convert to an error that we can report without crashing:

```suggestion
                    let Some(local_tab) = gui_tabs.get(&pane.window_id) else {
                       anyhow::bail!("invalid tmux window id {}", pane.window_id);
                    }
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
+            for l in &window.layout {
+                match l {
+                    TmuxLayout::SinglePane(p) => {
+                        let local_pane = self.create_pane(&p).unwrap();
```
bare unwrap here; in general, if you are in a function that returns a `Result` of some kind in wezterm, you should strive to propagate an error rather than producing a panic.  It's especially important when processing the results from a remote protocol because connection or version mismatching weirdness really shouldn't be able to kill wezterm and break its other sessions that could have critical state for the user!

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
+                for p in split_stack {
+                    let local_pane;
+                    if !self.check_pane_attached(p.window_id, p.pane_id) {
+                        local_pane = self.create_pane(&p).unwrap();
```
Can we propagate a reasonable error here instead of panic?
```suggestion
                        local_pane = self.create_pane(&p).context("failed to create pane")?;
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
+                        )? + 1;
+                    } else {
+                        let pane_map = self.remote_panes.lock();
+                        let local_pane_id = pane_map.get(&p.pane_id).unwrap().lock().local_pane_id;
```
unwrap -> error

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
-                .push_back(Box::new(CapturePane(pane.pane_id)));
-            TmuxDomainState::schedule_send_next_command(self.domain_id);
+            let gui_tabs = self.gui_tabs.lock();
+            let local_tab = gui_tabs.get(&window.window_id).unwrap();
```
unwrap -> error

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
+
+        let mut gui_tabs = tmux_domain.inner.gui_tabs.lock();
+
+        let local_tab = gui_tabs.get_mut(&self.window_id).unwrap();
```
defang this unwrap too!

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -237,12 +649,19 @@ impl TmuxCommand for ListAllPanes {
             let session_id = session_id[1..].parse()?;
             let window_id = window_id[1..].parse()?;
             let pane_id = pane_id[1..].parse()?;
+            let pane_active = if pane_active == 1 { true } else { false };
```
```suggestion
            let pane_active = pane_active == 1;
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -255,14 +674,267 @@ impl TmuxCommand for ListAllPanes {
+    }
+}
+
+fn parse_pane(layout: &str) -> anyhow::Result<PaneItem> {
```
all the unwraps and expects in here need to be converted to error propagation so that wezterm cannot die just because it got something it didn't understand from tmux!

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -255,14 +674,267 @@ impl TmuxCommand for ListAllPanes {
+}
+
+fn parse_pane(layout: &str) -> anyhow::Result<PaneItem> {
+    let re_pane = Regex::new(r"^,?(\d+)x(\d+),(\d+),(\d+)(,(\d+))?").unwrap();
```
With the Regex stuff, I'm a bit conflicted.  The intent of the `tmux.pest` file is that it encapsulates the parsing of the tmux protocol in a well-structured way that is a bit more rigorous than a bunch of regular expressions.

I am pragmatic though: if there is a good reason why it is not possible to express things in pest, then we can accommodate the regex.  If we need to go the regex route, then I'd recommend using a LazyLock to compile each regex once:

```rust
static RE_PANE: LazyLock<Regex> = LazyLock::new(|| Regex::new("something").unwrap());
```

(I'm cool with an unwrap in this situation: if the regex doesn't parse, that's a showstopping problem)

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -255,14 +674,267 @@ impl TmuxCommand for ListAllPanes {
+fn parse_layout(
+    mut layout: &str,
+    result: &mut Vec<TmuxLayout>,
+) -> anyhow::Result<(Option<TmuxLayout>, usize)> {
```
same commentary about unwraps and Regex in here!

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -294,8 +1036,12 @@ impl TmuxCommand for Resize {
+    fn get_command(&self, _domain_id: DomainId) -> String {
+        const HISTORY_LINES: isize = -2000;
+        format!(
+            "capture-pane -p -t %{} -e -C -S {}\n",
+            self.0, HISTORY_LINES
```
FWIW, you can slightly simplify this:

```suggestion
            "capture-pane -p -t %{} -e -C -S {HISTORY_LINES}\n",
            self.0
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -316,6 +1062,11 @@ impl TmuxCommand for CapturePane {
         let pane_map = tmux_domain.inner.remote_panes.lock();
         if let Some(pane) = pane_map.get(&self.0) {
             let mut pane = pane.lock();
+            match mux.get_pane(pane.local_pane_id) {
+                Some(p) => p.set_cursor_position(0, 0),
+                None => {}
+            }
```
```suggestion
            if let Some(p) = mux.get_pane(pane.local_pane_id) {
                p.set_cursor_position(0, 0);
            }
```

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -343,3 +1094,209 @@ impl TmuxCommand for SendKeys {
+        let mux = Mux::get();
+        let domain = match mux.get_domain(domain_id) {
+            Some(d) => d,
+            None => panic!("Tmux domain lost"),
```
panic -> anyhow::bail throughout!

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -343,3 +1094,209 @@ impl TmuxCommand for SendKeys {
+    use super::*;
+
+    #[test]
+    fn test_parse_pane() {
```
Note that my comments about unwraps -> errors don't apply to tests; those are a great place for a panic when checking assertions!

wez commented on termwiz/src/tmux_cc/mod.rs:
```diff
@@ -285,8 +255,8 @@ fn parse_line(line: &str) -> anyhow::Result<Event> {
             let window = parse_window_id(pairs.next().unwrap())?;
-            let layout = pairs.next().and_then(parse_window_layout).unwrap();
-            let visible_layout = pairs.next().and_then(parse_window_layout);
+            let layout = unvis(pairs.next().unwrap().as_str())?;
```
Can you defang these unwraps too? I realize that a bunch of these are inherited from the existing code

wez commented on mux/src/tmux_commands.rs:
```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
+        return Ok(pane);
+    }
+
+    pub fn fix_attached_pane_id(
```
Would be good to have some commentary here about what this is fixing and why!

wez commented on mux/src/tmux.rs:
```diff
@@ -206,6 +329,32 @@ impl TmuxDomain {
+    ) -> anyhow::Result<Arc<dyn Pane>> {
+        self.inner.split_tmux_pane(tab, pane_id, split_request);
+
+        // Give a fake id for now, and fix it later on event WindowPaneChanged
```
is there a risk that this fake id will cause the original pane to get broken or removed in the wrong circumstances later?

If so, then you could `alloc_pane_id()` a unique and unused id?

joexue commented:
> * pest

True, I'm removing the set_cursor_position methods and moving the parse_pane and parse_layout into tmux_cc to use pest.

joexue commented on mux/src/tmux_commands.rs:
```diff
@@ -294,8 +1036,12 @@ impl TmuxCommand for Resize {
-    fn get_command(&self) -> String {
-        format!("capturep -p -t %{} -e -C\n", self.0)
+    fn get_command(&self, _domain_id: DomainId) -> String {
+        const HISTORY_LINES: isize = -2000;
```
right, I thought it, we can put it off for now, I have a couple of plans also need the configuration, such as how to handle the bootstrap pane.

joexue commented:
I have addressed all comments, please continue to review it.

beside that, I changed the behavior a bit, that the we don't create a standalone window for tmux domain.
Instead, we use the same window which the bootstrap pane (the pane run "tmux -CC") in. Then user don't need to manager a extra window which maybe only bootstrap pane occupies.

If user really want a stand alone window separate the local tabs/panes, user can create a new wezterm instance, then start tmux from that instance. Anyway, put the tmux bootstrap pane with tmux tabs/panes are more reasonable then leave bootstrap pane with local tabs/panes. That is also my flavor when I use iTerm2. Of cause we can use configuration to control this behavior later.  (see fig below)

For history, I use a new approach to get the tmux history_limit instead of using the hard code 2000 lines.

iDarshan commented:
This latest update doesn't work on Mac OS 15.3. Doing ssh to a server and then running `$ tmux -CC attach -t wezterm` Prints below error.
```
15:56:18.472  INFO   mux::tmux          > tmux session changed:0
15:56:18.605  ERROR  mux::tmux_commands > Error resizing: domain_id=8 result=Guarded { error: true, timestamp: 1740009378, number: 33879, flags: 1, output: "can't find session 0\n" }
15:56:18.605  ERROR  mux::tmux          > Tmux processing command result error: list-windows failed

joexue commented:
> This latest update doesn't work on Mac OS 15.3. Doing ssh to a server and then running `$ tmux -CC attach -t wezterm` Prints below error.
> 
> ```
> 15:56:18.472  INFO   mux::tmux          > tmux session changed:0
> 15:56:18.605  ERROR  mux::tmux_commands > Error resizing: domain_id=8 result=Guarded { error: true, timestamp: 1740009378, number: 33879, flags: 1, output: "can't find session 0\n" }
> 15:56:18.605  ERROR  mux::tmux          > Tmux processing command result error: list-windows failed
> ```

can you set WEZTEMLOG=mux=debug,info and attach the log ?

iDarshan commented:
> > This latest update doesn't work on Mac OS 15.3. Doing ssh to a server and then running `$ tmux -CC attach -t wezterm` Prints below error.
> > ```
> > 15:56:18.472  INFO   mux::tmux          > tmux session changed:0
> > 15:56:18.605  ERROR  mux::tmux_commands > Error resizing: domain_id=8 result=Guarded { error: true, timestamp: 1740009378, number: 33879, flags: 1, output: "can't find session 0\n" }
> > 15:56:18.605  ERROR  mux::tmux          > Tmux processing command result error: list-windows failed
> > ```
> 
> can you set WEZTEMLOG=mux=debug,info and attach the log ? maybe your tmux doesn't support "list-window -t <session_id>"? could you help test it? just run this command "tmux list-window -t "?
> 
> Thanks

```
$ tmux list-window -t wezterm
0: bash#- (2 panes) [162x44] [layout f8f1,162x44,0,0{85x44,0,0,0,76x44,86,0,4}] @0
1: bash* (2 panes) [162x44] [layout 78f5,162x44,0,0{85x44,0,0,1,76x44,86,0,3}] @1 (active)
2: bash# (2 panes) [162x44] [layout 6f1a,162x44,0,0{84x44,0,0,2,77x44,85,0,5}] @2
3: bash# (2 panes) [162x44] [layout eeef,162x44,0,0{84x44,0,0,7,77x44,85,0,8}] @3
4: bash# (1 panes) [162x44] [layout bd06,162x44,0,0,9] @4
$

joexue commented:
> > > This latest update doesn't work on Mac OS 15.3. Doing ssh to a server and then running `$ tmux -CC attach -t wezterm` Prints below error.
> > > ```
> > > 15:56:18.472  INFO   mux::tmux          > tmux session changed:0
> > > 15:56:18.605  ERROR  mux::tmux_commands > Error resizing: domain_id=8 result=Guarded { error: true, timestamp: 1740009378, number: 33879, flags: 1, output: "can't find session 0\n" }
> > > 15:56:18.605  ERROR  mux::tmux          > Tmux processing command result error: list-windows failed
> > > ```
> > 
> > 
> > can you set WEZTEMLOG=mux=debug,info and attach the log ? maybe your tmux doesn't support "list-window -t <session_id>"? could you help test it? just run this command "tmux list-window -t "?
> > Thanks
> 
> ```
> $ tmux list-window -t wezterm
> 0: bash#- (2 panes) [162x44] [layout f8f1,162x44,0,0{85x44,0,0,0,76x44,86,0,4}] @0
> 1: bash* (2 panes) [162x44] [layout 78f5,162x44,0,0{85x44,0,0,1,76x44,86,0,3}] @1 (active)
> 2: bash# (2 panes) [162x44] [layout 6f1a,162x44,0,0{84x44,0,0,2,77x44,85,0,5}] @2
> 3: bash# (2 panes) [162x44] [layout eeef,162x44,0,0{84x44,0,0,7,77x44,85,0,8}] @3
> 4: bash# (1 panes) [162x44] [layout bd06,162x44,0,0,9] @4
> $
> ```

could you use id instead of name?

iDarshan commented:
> please try the latest, looks your gave your session name make the session name and session id don't match, we use session id as the parameter. in this case , we should be more strict, give an id prefix $.
> 
> Thanks. @iDarshan

It works now! But I still see these errors:
```
19:16:06.206  INFO   mux::localpane > tmux -CC mode requested
19:16:06.206  INFO   mux::tmux      > tmux session changed:0
19:16:06.206  ERROR  mux::tmux      > Tmux pane 1 havn't been attached
19:16:06.206  ERROR  mux::tmux      > Tmux pane 2 havn't been attached
19:16:06.206  ERROR  mux::tmux      > Tmux pane 3 havn't been attached
19:16:06.206  ERROR  mux::tmux      > Tmux pane 5 havn't been attached
19:16:06.302  INFO   mux::tmux      > Tmux create window id 0

joexue commented:
> > please try the latest, looks your gave your session name make the session name and session id don't match, we use session id as the parameter. in this case , we should be more strict, give an id prefix $.
> > Thanks. @iDarshan
> 
> It works now! But I still see these errors:
> 
> ```
> 19:16:06.206  INFO   mux::localpane > tmux -CC mode requested
> 19:16:06.206  INFO   mux::tmux      > tmux session changed:0
> 19:16:06.206  ERROR  mux::tmux      > Tmux pane 1 havn't been attached
> 19:16:06.206  ERROR  mux::tmux      > Tmux pane 2 havn't been attached
> 19:16:06.206  ERROR  mux::tmux      > Tmux pane 3 havn't been attached
> 19:16:06.206  ERROR  mux::tmux      > Tmux pane 5 havn't been attached
> 19:16:06.302  INFO   mux::tmux      > Tmux create window id 0
> ```

Those are ok, we handled those cases

bfstree commented:
https://github.com/user-attachments/assets/35e20687-6c3b-4b9a-b82d-be864ff9977f

`lazygit` is not working properly in tmux CC mode. I am receiving the following errors:

```
18:11:17.356  INFO   mux::localpane > tmux -CC mode requested
18:11:17.357  INFO   mux::tmux      > Tmux create window id 1
18:11:17.357  INFO   mux::tmux      > tmux session changed:0
18:11:17.417  INFO   mux::tmux_commands > new pane synced, id: 0
18:11:39.375  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1043
18:11:39.376  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
18:11:40.221  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1255
18:11:40.221  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
18:11:43.500  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1457
18:11:43.500  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
18:11:50.287  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1523
18:11:50.288  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
```

This is on macOS 15.3, tmux 3.5a, lazygit 0.46, TERM=tmux-256color, and using the latest wezterm commit of this branch.
This does not happen in normal tmux mode.

Thank you for implementing this feature!

joexue commented:
> wezterm-lazygit.mp4 
> 
> `lazygit` is not working properly in tmux CC mode. I am receiving the following errors:
> 
> ```
> 18:11:17.356  INFO   mux::localpane > tmux -CC mode requested
> 18:11:17.357  INFO   mux::tmux      > Tmux create window id 1
> 18:11:17.357  INFO   mux::tmux      > tmux session changed:0
> 18:11:17.417  INFO   mux::tmux_commands > new pane synced, id: 0
> 18:11:39.375  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1043
> 18:11:39.376  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
> 18:11:40.221  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1255
> 18:11:40.221  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
> 18:11:43.500  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1457
> 18:11:43.500  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
> 18:11:50.287  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: incomplete utf-8 byte sequence from index 1523
> 18:11:50.288  ERROR  termwiz::tmux_cc   > Failed to parse line from tmux: invalid utf-8 sequence of 1 bytes from index 11
> ```
> 
> This is on macOS 15.3, tmux 3.5a, lazygit 0.46, TERM=tmux-256color, and using the latest wezterm commit of this branch. This does not happen in normal tmux mode.
> 
> Thank you for implementing this feature!

@wez Should we change the from_utf8 to from_utf8_lossy in tmux_cc crate? Since the APP may output non-UTF8 code by design.

wez commented:
Why is it non-utf8? is that because the app is doing that, or because tmux's protocol is doing that?

If it is the app: yeah, let's use a lossy decode in that case.
If it is tmux outputting non-utf8, can we force it to be utf8? What's happening there?

wez commented on termwiz/src/tmux_cc/mod.rs:
```diff
@@ -716,4 +829,58 @@ here
+        let layout_case2 = "158x40,0,0[158x20,0,0,69,158x19,0,21{79x19,0,21,70,78x19,80,21[78x9,80,21,71,78x9,80,31,73]}]".to_string();
+        let layout_case3 = "158x40,0,0{79x40,0,0[79x20,0,0,74,79x19,0,21{39x19,0,21,76,39x19,40,21,77}],78x40,80,0,75}".to_string();
+
+        let mut layout = parse_layout(layout_case1.get(0..).unwrap()).unwrap();
```
```suggestion
        let mut layout = parse_layout(&layout_case1).unwrap();
```

wez commented on termwiz/src/tmux_cc/mod.rs:
```diff
@@ -716,4 +829,58 @@ here
+            false
+        });
+
+        layout = parse_layout(layout_case2.get(0..).unwrap()).unwrap();
```
```suggestion
        layout = parse_layout(&layout_case2).unwrap();
```

wez commented on termwiz/src/tmux_cc/mod.rs:
```diff
@@ -716,4 +829,58 @@ here
+            false
+        });
+
+        layout = parse_layout(layout_case3.get(0..).unwrap()).unwrap();
```
```suggestion
        layout = parse_layout(&layout_case3).unwrap();
```

wez commented on termwiz/src/tmux_cc/mod.rs:
```diff
@@ -716,4 +829,58 @@ here
+        });
+
+        layout = parse_layout(layout_case2.get(0..).unwrap()).unwrap();
+        assert!(if let WindowLayout::SplitVertical(_x) = &layout[0] {
+            true
+        } else {
+            false
+        });
```
you can rewrite these like this:

```suggestion
        assert!(matches!(&layout[0], WindowLayout::SplitVertical(_x)));
```

wez commented on termwiz/src/tmux_cc/mod.rs:
```diff
@@ -265,17 +281,7 @@ fn parse_line(line: &str) -> anyhow::Result<Event> {
-        | Rule::line_entire
-        | Rule::EOI
-        | Rule::number => unreachable!(),
+        _ => unreachable!(),
```
could also consider swapping unreachable for a bail! here, for extra safety

joexue commented:
> Why is it non-utf8? is that because the app is doing that, or because tmux's protocol is doing that?
> 
> If it is the app: yeah, let's use a lossy decode in that case. If it is tmux outputting non-utf8, can we force it to be utf8? What's happening there?

from my understanding, it is the app

joexue commented:
@wez , I have addressed your second round comments.
Thank you for your reviewing and very detailed change suggestions. I learnt lots lots from this.  

For pending_split you suggested, I use VecDeq for it, since when we act split we have the original split pane information, and when the new pane generated, the tmux only gives new pane id, so we cannot use a HashMap. But out command is in order, hence the tmux action/event is in order too, so the VecDeq works for this situation.

iDarshan commented:
@joexue Wezterm hangs when ssh to server, with which tmux control mode is running, disconnected or times out. Need to Force quit Wezterm in such situation.

```
  |
1 | Read from remote host test-dev: Operation timed out
  | ^---
  |
  = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change
13:08:39.459  INFO   mux                > domain detached panes: [1, 3, 4, 5, 2]

joexue commented on mux/src/tmux_commands.rs:
```diff
@@ -255,14 +674,267 @@ impl TmuxCommand for ListAllPanes {
+#[derive(Debug)]
+pub(crate) struct ListAllWindows {
+    pub session_id: TmuxSessionId,
+    pub window_id: Option<TmuxWindowId>,
```
do you mean window_id? yes, it is being used.

iDarshan commented:
> > @joexue Wezterm hangs when ssh to server, with which tmux control mode is running, disconnected or times out. Need to Force quit Wezterm in such situation.
> 
> > 
> 
> > ```
> 
> >   |
> 
> > 1 | Read from remote host test-dev: Operation timed out
> 
> >   | ^---
> 
> >   |
> 
> >   = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change
> 
> > 13:08:39.459  INFO   mux                > domain detached panes: [1, 3, 4, 5, 2]
> 
> > ```
> 
> 
> 
> can you give more details? thanks

What kind of details? This is basic. Ssh to server, run tmux -CC attach. Leave the session idle, ssh will timeout and wezterm will hang. Mac OS 15.3.

joexue commented:
> > > @joexue Wezterm hangs when ssh to server, with which tmux control mode is running, disconnected or times out. Need to Force quit Wezterm in such situation.
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > |
> > 
> > 
> > > 1 | Read from remote host test-dev: Operation timed out
> > 
> > 
> > > | ^---
> > 
> > 
> > > |
> > 
> > 
> > > = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change
> > 
> > 
> > > 13:08:39.459  INFO   mux                > domain detached panes: [1, 3, 4, 5, 2]
> > 
> > 
> > > 
> > 
> > 
> > can you give more details? thanks
> 
> What kind of details? This is basic. Ssh to server, run tmux -CC attach. Leave the session idle, ssh will timeout and wezterm will hang. Mac OS 15.3.

I cannot reproduce it here(not MacOS), once the ssh server disconnect connection , tmux cc mode will end and ssh connection too. Is that your ssh behaviour? Try to just ssh to host then let it timeout.

joexue commented:
> > > > > @joexue Wezterm hangs when ssh to server, with which tmux control mode is running, disconnected or times out. Need to Force quit Wezterm in such situation.
> > > > 
> > > > 
> > > > > 
> > > > 
> > > > 
> > > > > 
> > > > 
> > > > 
> > > > > |
> > > > 
> > > > 
> > > > > 1 | Read from remote host test-dev: Operation timed out
> > > > 
> > > > 
> > > > > | ^---
> > > > 
> > > > 
> > > > > |
> > > > 
> > > > 
> > > > > = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change
> > > > 
> > > > 
> > > > > 13:08:39.459  INFO   mux                > domain detached panes: [1, 3, 4, 5, 2]
> > > > 
> > > > 
> > > > > 
> > > > 
> > > > 
> > > > can you give more details? thanks
> > > 
> > > 
> > > What kind of details? This is basic. Ssh to server, run tmux -CC attach. Leave the session idle, ssh will timeout and wezterm will hang. Mac OS 15.3.
> > 
> > 
> > I cannot reproduce it here(not MacOS), once the ssh server disconnect connection , tmux cc mode will end and ssh connection too. Is that your ssh behaviour? Try to just ssh to host then let it timeout.
> 
> No, ssh is fine. Also, control mode works fine in iTerm2 on mac os when ssh times out, it safely gets detached.

hi @iDarshan could you help to try the below diff?
```
diff --git a/mux/src/lib.rs b/mux/src/lib.rs
index f46e5ee08..59142094a 100644
--- a/mux/src/lib.rs
+++ b/mux/src/lib.rs
@@ -1069,6 +1069,7 @@ impl Mux {
     }

     pub fn domain_was_detached(&self, domain: DomainId) {
+/*
         let mut dead_panes = vec![];
         for pane in self.panes.read().values() {
             if pane.domain_id() == domain {
@@ -1076,6 +1077,7 @@ impl Mux {
             }
         }

+*/
         {
             let mut windows = self.windows.write();
             for (_, win) in windows.iter_mut() {
@@ -1085,11 +1087,13 @@ impl Mux {
             }
         }

+/*
         log::info!("domain detached panes: {:?}", dead_panes);
         for pane_id in dead_panes {
             self.remove_pane_internal(pane_id);
         }

+*/
         self.prune_dead_windows();
     }
```

iDarshan commented:
> > > > > > @joexue Wezterm hangs when ssh to server, with which tmux control mode is running, disconnected or times out. Need to Force quit Wezterm in such situation.
> > > > > 
> > > > > 
> > > > > > 
> > > > > 
> > > > > 
> > > > > > 
> > > > > 
> > > > > 
> > > > > > |
> > > > > 
> > > > > 
> > > > > > 1 | Read from remote host test-dev: Operation timed out
> > > > > 
> > > > > 
> > > > > > | ^---
> > > > > 
> > > > > 
> > > > > > |
> > > > > 
> > > > > 
> > > > > > = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change
> > > > > 
> > > > > 
> > > > > > 13:08:39.459  INFO   mux                > domain detached panes: [1, 3, 4, 5, 2]
> > > > > 
> > > > > 
> > > > > > 
> > > > > 
> > > > > 
> > > > > can you give more details? thanks
> > > > 
> > > > 
> > > > What kind of details? This is basic. Ssh to server, run tmux -CC attach. Leave the session idle, ssh will timeout and wezterm will hang. Mac OS 15.3.
> > > 
> > > 
> > > I cannot reproduce it here(not MacOS), once the ssh server disconnect connection , tmux cc mode will end and ssh connection too. Is that your ssh behaviour? Try to just ssh to host then let it timeout.
> > 
> > 
> > No, ssh is fine. Also, control mode works fine in iTerm2 on mac os when ssh times out, it safely gets detached.
> 
> hi @iDarshan could you help to try the below diff?
> 
> ```
> diff --git a/mux/src/lib.rs b/mux/src/lib.rs
> index f46e5ee08..59142094a 100644
> --- a/mux/src/lib.rs
> +++ b/mux/src/lib.rs
> @@ -1069,6 +1069,7 @@ impl Mux {
>      }
> 
>      pub fn domain_was_detached(&self, domain: DomainId) {
> +/*
>          let mut dead_panes = vec![];
>          for pane in self.panes.read().values() {
>              if pane.domain_id() == domain {
> @@ -1076,6 +1077,7 @@ impl Mux {
>              }
>          }
> 
> +*/
>          {
>              let mut windows = self.windows.write();
>              for (_, win) in windows.iter_mut() {
> @@ -1085,11 +1087,13 @@ impl Mux {
>              }
>          }
> 
> +/*
>          log::info!("domain detached panes: {:?}", dead_panes);
>          for pane_id in dead_panes {
>              self.remove_pane_internal(pane_id);
>          }
> 
> +*/
>          self.prune_dead_windows();
>      }
> ```

This patch didn't help. It still gets hanged. Error message on console:
```
14:56:59.240  ERROR  termwiz::tmux_cc   > Unrecognized tmux cc line:  --> 1:1
  |
1 | Read from remote host test-dev: Operation timed out
  | ^---
  |
  = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change

joexue commented:
> > > > > > > @joexue Wezterm hangs when ssh to server, with which tmux control mode is running, disconnected or times out. Need to Force quit Wezterm in such situation.
> > > > > > 
> > > > > > 
> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > > |
> > > > > > 
> > > > > > 
> > > > > > > 1 | Read from remote host test-dev: Operation timed out
> > > > > > 
> > > > > > 
> > > > > > > | ^---
> > > > > > 
> > > > > > 
> > > > > > > |
> > > > > > 
> > > > > > 
> > > > > > > = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change
> > > > > > 
> > > > > > 
> > > > > > > 13:08:39.459  INFO   mux                > domain detached panes: [1, 3, 4, 5, 2]
> > > > > > 
> > > > > > 
> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > can you give more details? thanks
> > > > > 
> > > > > 
> > > > > What kind of details? This is basic. Ssh to server, run tmux -CC attach. Leave the session idle, ssh will timeout and wezterm will hang. Mac OS 15.3.
> > > > 
> > > > 
> > > > I cannot reproduce it here(not MacOS), once the ssh server disconnect connection , tmux cc mode will end and ssh connection too. Is that your ssh behaviour? Try to just ssh to host then let it timeout.
> > > 
> > > 
> > > No, ssh is fine. Also, control mode works fine in iTerm2 on mac os when ssh times out, it safely gets detached.
> > 
> > 
> > hi @iDarshan could you help to try the below diff?
> > ```
> > diff --git a/mux/src/lib.rs b/mux/src/lib.rs
> > index f46e5ee08..59142094a 100644
> > --- a/mux/src/lib.rs
> > +++ b/mux/src/lib.rs
> > @@ -1069,6 +1069,7 @@ impl Mux {
> >      }
> > 
> >      pub fn domain_was_detached(&self, domain: DomainId) {
> > +/*
> >          let mut dead_panes = vec![];
> >          for pane in self.panes.read().values() {
> >              if pane.domain_id() == domain {
> > @@ -1076,6 +1077,7 @@ impl Mux {
> >              }
> >          }
> > 
> > +*/
> >          {
> >              let mut windows = self.windows.write();
> >              for (_, win) in windows.iter_mut() {
> > @@ -1085,11 +1087,13 @@ impl Mux {
> >              }
> >          }
> > 
> > +/*
> >          log::info!("domain detached panes: {:?}", dead_panes);
> >          for pane_id in dead_panes {
> >              self.remove_pane_internal(pane_id);
> >          }
> > 
> > +*/
> >          self.prune_dead_windows();
> >      }
> > ```
> 
> This patch didn't help. It still gets hanged. Error message on console:
> 
> ```
> 14:56:59.240  ERROR  termwiz::tmux_cc   > Unrecognized tmux cc line:  --> 1:1
>   |
> 1 | Read from remote host test-dev: Operation timed out
>   | ^---
>   |
>   = expected begin, end, error, client_session_changed, client_detached, output, exit, sessions_changed, pane_mode_changed, window_add, window_close, window_pane_changed, window_renamed, session_changed, session_renamed, session_window_changed, or layout_change
> ```

Could you help to try below? I found dead lock in infra code. The diff is from cellphone, the format may not very correct, please change it accordingly.
<pre>

diff --git a/mux/src/lib.rs b/mux/src/lib.rs
index f465ee08.. a177bb841 100644
--- a/mux/src/lib.rs
*+ b/mux/src/lib.rs
@@ - 807,9 +807,9 @@ impl Mux {
        if Let Some(pane) = self.panes write().remove(&pane_id) .clone() {
        Log:: debug! ("killing pane f}", pane_id);                 
        pane.kill();
-       self.recompute_pane_count();
self.notify(MuxNotification::PaneRemoved(pane_id));
          }
+       self.recompute_pane_count();
    }

    fn remove_tab_internal(&self, tab_id: TabId) → Option<Arc<Tab»> {

</pre>

iDarshan commented:
> self.recompute_pane_count

This patch works! Doesn't hang anymore on ssh timeout, safely gets detached. Thanks!

joexue commented:
> > self.recompute_pane_count
> 
> This patch works! Doesn't hang anymore on ssh timeout, safely gets detached. Thanks!

Thank you for your testing 
In this case it should no be in this scope, do you mind raising an issue and move the valid context there? I can put my investigation there too. Let @wez make decision how to fix it. My patch is just  a workaround.

iDarshan commented:
> > > self.recompute_pane_count
> > 
> > 
> > This patch works! Doesn't hang anymore on ssh timeout, safely gets detached. Thanks!
> 
> Thank you for your testing In this case it should not be in this scope, do you mind raising an issue and move the valid context there? I can put my investigation there too. Let @wez make decision how to fix it. My patch is just a workaround.
> 
> Or let me know if you prefer I do it.

- Sure, I have created issue #6749

joexue commented:
> Wezterm crashed as soon as I tried to run it with `tmux -CC a -t my-session-name`.
> 
> Steps:
> 
> 1. `cargo run --release --bin wezterm -- start` on Fedora 41 + Window Maker in a virtual machine on a M2 MBP.
> 2. et (Eternal Terminal) to my remote Linux box running Debian.
> 3. `tmux -CC a -t my-session-name`
> 
> Logs attached: [wezterm.log](https://github.com/user-attachments/files/19099866/wezterm.log)

The log shows you are not running this PR version, the crash issue you met actually a known one and fixed by this PR. Thanks

felixding commented:
Oops, my bad.

I've successfully compiled and run WezTerm using your branch. So far so good. Thanks!

wez commented:
Thank you very much for this!

justrajdeep commented:
Cant believe my luck ... today only i came to learn about control mode and today it got implemented in wezterm 😁
Does anyone know when it will get reflected in brew packages?

joexue commented:
> Thank you very much for this!

Thank you @wez , I learnt much from this project, especially from your code review suggestions.

felixding commented:
Found a bug that I can reliably reproduce:

1. `tmux -CC a -t my-session-name`
2. `ssh server-ip`
3. `tmux`

Then, this line popped up:

![Screenshot 2025-03-10 at 11 05 55](https://github.com/user-attachments/assets/66523522-12eb-4556-acb9-43f8e18abe7b)

wez commented:
If you find issues with the implementation that is now in main, please file a new issue and cc @joexue there, so that this PR doesn't become overwhelmed. Thanks!
